sdk/lib/opentelemetry/sdk/trace/export/batch_sampled_span_processor.rb

```diff
@@ -21,7 +22,92 @@ module Export
         # worker thread that exports the spans to wake up and start a new
         # export cycle.
         class BatchSampledSpanProcessor
-          # TODO
+          SCHEDULE_DELAY_MILLIS = 5000
+          MAX_QUEUE_SIZE = 2048
+          MAX_EXPORT_BATCH_SIZE = 512
+
+          def initialize(exporter:, schedule_delay: SCHEDULE_DELAY_MILLIS, max_queue_size: MAX_QUEUE_SIZE, max_export_batch_size: MAX_EXPORT_BATCH_SIZE)
+            @exporter = exporter
+            @mutex = Mutex.new
+            @condition = ConditionVariable.new
+            @keep_running = true
+            @delay = schedule_delay
+            @max_queue_size = max_queue_size
+            @batch_size = max_export_batch_size
+            @spans = []
+            @thread = Thread.new { work }
+          end
+
+          # does nothing for this processor
+          def on_start(span)
+            # noop
+          end
+
+          # adds a span to the batcher, threadsafe may block on lock
+          def on_end(span)
+            lock do
+              spans.shift if spans.size >= max_queue_size
+              spans << span
+              @condition.signal if spans.size > max_queue_size/2
+            end
+          end
+
+          # shuts the consumer thread down and flushes the current accumulated buffer
+          # will block until the thread is finished
+          def shutdown
+            lock do
+              @keep_running = false
+              @condition.signal
+            end
+
+            @thread.join
+          end
+
+          private
+
+          attr_reader :spans, :max_queue_size, :batch_size
+
+          def work
+            loop do
+              keep_running = nil
+              batch = nil
+              lock do
+                if  spans.size < max_queue_size
+                  loop do
+                    break if !spans.empty? || !@keep_running
+                    @condition.wait(@mutex, @delay)
+                  end
+                end
+                keep_running = @keep_running
```

Comment:
Why not `break unless @keep_running` here and avoid the local var? Then rely on `flush` in the `shutdown` to do the final exports?